code/game/objects/items/breakable.dm

```diff
@@ -0,0 +1,260 @@
+	var/breaks_text		//Visible message when the items breaks. eg. "breaks apart"
+	//Sounds:
+	var/breaks_sound	//Path to audible sound when the item breaks apart.
+	var/damaged_sound	//Path to audible sound when the item is damaged but not fully destroyed.
+	var/glanced_sound	//Path to audible sound when the item recives a glancing attack not strong enough to damage it.
+
+/obj/item/proc/on_broken(var/atom/target, var/range, var/speed, var/override, var/fly_speed, var/atom/hit_atom) //Called right before an object breaks.
+	//Drop and and propel any fragments:
+	drop_fragments(target,range,speed,override,fly_speed)
+	//Drop and propel any contents:
+	drop_contents(target,range,speed,override,fly_speed)
+	//Spill any reagents:
+	spill_reagents(hit_atom)
+	if(breaks_text)
+		visible_message("<span class='warning'>\The [src] [breaks_text]!</span>")
+	if(breaks_sound)
+		playsound(src, breaks_sound, 50, 1)
+	else if(damaged_sound)
+		playsound(src, damaged_sound, 50, 1)
+
+/obj/item/proc/drop_fragments(var/atom/target, var/range, var/speed, var/override, var/fly_speed) //Separate proc in case special stuff happens with a given item's fragments. Parameters are for throwing the fragments.
+	if(breakable_fragments.len)
+		var/oneeach=(isnull(fragment_amounts) || breakable_fragments.len != fragment_amounts.len) //default to 1 of each fragment type if fragment_amounts isn't specified or there's a length mismatch
+		var/numtodrop
+		var/thisfragment
+		for(var/frag_ind in 1 to breakable_fragments.len)
+			if(oneeach)
+				numtodrop=1
+			else
+				numtodrop=fragment_amounts[frag_ind]
+			thisfragment=breakable_fragments[frag_ind]
+			for(var/n in 1 to numtodrop)
+				var/obj/item/O = new thisfragment (get_turf(src))
+				//Transfer fingerprints, fibers, and bloodstains to the fragment.
+				transfer_fingerprints(src,O)
+				transfer_item_blood_data(src,O)
+				if(target && range && speed) //Propel the fragment if specified.
+					O.throw_at(target,range,speed,override,fly_speed)
+
+/obj/item/proc/drop_contents(var/atom/target, var/range, var/speed, var/override, var/fly_speed) //Drop the contents of the item and propel them if the item itself received a propulsive blow.
+	if(contents.len)
+		for(var/obj/item/thiscontent in contents)
+			thiscontent.forceMove(src.loc)
+			if(target && range && speed) //Propel the content if specified.
+				thiscontent.throw_at(target,range,speed,override,fly_speed)
+
+/obj/item/proc/spill_reagents(var/atom/hit_atom) //Spill any reagents contained within the item onto the floor, and the atom it hit when it broke, if applicable.
+	if(!isnull(reagents))
+		if(!isnull(hit_atom) && hit_atom != get_turf(src)) //If it hit something other than the floor, spill it onto that.
+			reagents.reaction(hit_atom, TOUCH)
+		reagents.reaction(get_turf(src), TOUCH) //Then spill it onto the floor.
+
+/obj/item/proc/take_damage(var/incoming_damage, var/mute = TRUE)
```

Comment:
It's very strange that `take_damage()` would not also handle breaking the item if its health drops to/below zero. I see the places where there's code between the `take_damage()` and `break_item()` but I can't imagine there's not a better way to handle that.

Reply:
One reason is that the text generation for various types of attacks needs to know in advance what happened to the item when it was attacked. If it happens out of order, then the message about attacking the item will occur after the item breaks, leading to things like:
> You attack the breaker with the crowbar, cracking it!
> The beaker shatters!
> You attack the beaker with the crowbar!

To say the attack text within `take_damage()` or `break_item()` would require each new type of attack or damage source to be accounted for within those procs, or a flag set on trying to break the item or not in each situation. Whichever way, its somewhat complicated but keeping the functionality of the procs simple and separate is why this way seems to me to be optimal. But maybe there's a good solution?

Reply:
The cases where there's actual behavior between the two procs are pretty unavoidable. They do have to be called separately for those. The cases with just messages *could* be handled inside the procs, which would be simple and clean if we had real format strings, but we don't. If it weren't for the cases that required them to be called separately, I'd say handle them inside anyway, but since they have to be separated sometimes anyway it's fine.
However, none of this stops you from calling `break_item()` inside `take_damage()` by default and having an arg that tells it not to, or wrapping them both in a different proc to be used when possible, or something. Remember that other people work on the codebase, and having to separately call two procs in all cases just because they have to be separated in a few isn't very intuitive.